spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,173 @@
+		Assert.notNull(defaultType, "'defaultType' must not be null");
+		this.defaultType = defaultType;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 * @return the transformer
+	 */
+	public FromProtobufTransformer typeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+		return this;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 * @return the transformer
+	 */
+	public FromProtobufTransformer typeExpression(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+		return this;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpressionString(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	private void assertExpressionNotNull(Object expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+		Assert.state(message.getPayload() instanceof byte[], "Payload must be a byte[]");
+		Class<? extends com.google.protobuf.Message> type = null;
+		Object value = this.typeIdExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			type = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				type = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (type == null) {
+			type = this.defaultType;
+		}
+
+		try {
+			Builder messageBuilder = this.getMessageBuilder(type);
+			byte[] payload = (byte[]) message.getPayload();
+			com.google.protobuf.Message boza = messageBuilder.mergeFrom(payload).build();
+			return boza;
```

Comment:
I'm not sure in word `boza` and seems for me this variable can be eliminated altogether as a simple `return messageBuilder.mergeFrom(payload).build()`.
Although this might be for now: we might to do some post-processing eventually.
But still `boza` is not what I'd like to see in the production code 😄